### PR TITLE
[data][feat] add large dataset support

### PR DESCRIPTION
**Pull request by SwordFaith**
Thank you for the efforts on https://github.com/radixark/miles/pull/246 and the hard work contributed by @Ratish1!

### CODE REVIEW

**Comment by ChangyiYang** [comment]
@codex review

**Comment by ChangyiYang** [comment]
@codex review

File: slime/backends/fsdp_utils/actor.py [hunk 1]
```diff
@@ -712,7 +712,22 @@ def _has_rollout_log_probs(batch) -> bool:
...
 
-        loss = pg_loss - self.args.entropy_coef * entropy_loss
+        # Compute loss based on loss_type
+        if self.args.loss_type == "sft_loss":
+            # SFT loss: simple negative log-likelihood
+            sft_loss = -log_probs
+            if self.args.calculate_per_token_loss:
+                sft_loss = sum_of_token(sft_loss, response_lengths, loss_masks)
+            else:
+                sft_loss = sum_of_sample_mean(sft_loss, response_lengths, loss_masks)
```
Comment (PopSoda2002) [inline slime/backends/fsdp_utils/actor.py]: I think we just need to `calculate_per_sample_loss` in SFT?

**Review by PopSoda2002 — COMMENTED** [review]
Thanks for your great work! I am not sure if we need to seperate `datasets` and `SFT` into different PRs, currently this PR seems too large for reviewing

File: slime/rollout/data_source.py [hunk 2]
```diff
@@ -52,52 +52,157 @@ def __init__(self, args):
...
+                max_length=self.args.rollout_max_prompt_len,
+                prompt_key=self.args.input_key,
+                label_key=self.args.label_key,
+                tool_key=self.args.tool_key,
+                metadata_key=self.args.metadata_key,
+                multimodal_keys=self.args.multimodal_keys,
+                seed=self.args.rollout_seed,
+                apply_chat_template=self.args.apply_chat_template,
+                apply_chat_template_kwargs=self.args.apply_chat_template_kwargs,
+                dp_size=self._dp_size or 1,
```
Comment (zhuzilin) [inline slime/rollout/data_source.py]: We should not pass dp size to the data source because the data source is used in rollout manager, which does not have dp ranks.

**Comment by SwordFaith** [comment]
> Thanks for your great work! I am not sure if we need to seperate `datasets` and `SFT` into different PRs, currently this PR seems too large for reviewing

I’ll work on some simplifications with @ChangyiYang today. Afterward, could you review it again and discuss if we should split this PR into two?

File: slime/rollout/data_source.py (hunk 2 above)
Comment (SwordFaith) [reply slime/rollout/data_source.py]: > We should not pass dp size to the data source because the data source is used in rollout manager, which does not have dp ranks.

Apologies for the mistake, I’ll address it in the upcoming commits.

**Comment by PopSoda2002** [comment]
> > Thanks for your great work! I am not sure if we need to seperate `datasets` and `SFT` into different PRs, currently this PR seems too large for reviewing
> 
> I’ll work on some simplifications with @ChangyiYang today. Afterward, could you review it again and discuss if we should split this PR into two?

Yeah sure, definitely, always willing for help

File: slime/backends/fsdp_utils/actor.py (hunk 1 above)
Comment (ChangyiYang) [reply slime/backends/fsdp_utils/actor.py]: After further discussion, we realized that SFT should only use per-token loss.
We’ll simplify this logic by keeping token loss only here.
For users who still try to use sequence / per-sample loss in SFT, we’ll explicitly raise an error to avoid silent misconfiguration.
